File: January 2023/15 January 2023/Count the Substring.py
class Solution():
    def countSubstring(self, S):
        n= len(S)
        ans=0
        zero=n
        minus=0
        mp = [0]*(2*n+5)
        cur=zero
        for i in S:
            if(i=='0'):
                cur-=1
            else:
                cur+=1
            if(cur<=zero):
                minus+=1
            
            mp[cur]+=1
        for i in range(n):
            ans+=(n-i-minus)
            if(S[i]=='1'):
                mp[zero+1]-=1
                zero+=1
                minus+=mp[zero]
            
            else:
                mp[zero-1]-=1
                zero-=1
                minus-=1
                minus-=mp[zero+1]
            
        return ans

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == "__main__":
    for _ in range(int(input())):
        s = input()
        print(Solution().countSubstring(s))
# } Driver Code Ends

File: January 2023/15 January 2023/Count the Substring.cpp
//{ Driver Code Starts
//Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution{
public:
  long long countSubstring(string S){
    int n=S.size();
    long long ans=0;
    int sum=0,onesekam=0;
    unordered_map<int,int>mp;
    for(auto it:S){
        if(it=='1')sum++;
        else sum--;
        if(sum<=0)onesekam++;
        mp[sum]++;
    }
    sum=0;
    for(int i=0;i<n;i++){
        ans+=(n-i-onesekam);
        if(S[i]=='1'){
            sum++;
            mp[sum]--;
            onesekam+=mp[sum];
        }
        else{
            sum--;
            mp[sum]--;
            onesekam--;
            onesekam-=mp[sum+1];
        }
    }
    return ans;
  }
};

//{ Driver Code Starts.
int main() {
 ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
 int t=1;
 cin>>t;
 for(int i=1;i<=t;i++){
    string S;
    cin>>S;
    Solution obj;
    long long ans =obj.countSubstring(S);
    cout<<ans<<endl;
 }
}
// } Driver Code Ends
